### ISSUE TITLE
No visible python-39-c9s images on quay.io

### ISSUE THREAD

**Issue by michael131468** [sclorg/s2i-python-container#543]
The README of this project links to a quay.io repository space for CentOS Stream 9 based images:

https://quay.io/repository/sclorg/python-39-c9s

But (at least for me) there seems to be no image tags in this repo to use.

Is this expected?

**Comment by pkubatrh** [comment]
Hi @michael131468, thanks for spotting this! This is a leftover from before RHEL9 release, when we only had Centos Stream 9 content available. As the Python 3.9 is now available as an UBI image, it no longer makes sense to maintain Centos Stream 9 images, so the repository (and references to it) will be removed.

Just wondering, is there some use case that the UBI images do not cover for you and the Centos Stream 9 images do?

**Comment by guimou** [comment]
Hi @pkubatrh! Would you reconsider this decision?
Because as you mentioned, there are definitely use cases where UBI images don't cover all use cases. For example, installing R on UBI images is not possible, as required libraries are not available in UBI repos or Epel.
I have other use cases, almost all related to fancy data science libraries than we want to brinh to Open Data Hub or OpenShift Data Science.
Up until now I had been maintaining my own builb pipelines to create those c9s images (basically mimicking everything done there), but if I can have them from a more official repo, that would be best.

**Comment by guimou** [comment]
Hi @frenzymadness ! Sure thing.
I have created two log files from two of the images I'm building to use with Open Data Hub, but instead of starting from c9s, I'm starting from ubi9 (I extracted only the relevant parts/commands).
- One is fairly mainstream, it's just to create an R Kernel/RStudio deployment: 
[ubi9-opencv-pytorch.log](https://github.com/sclorg/s2i-python-container/files/10211086/ubi9-opencv-pytorch.log)
Main missing packages would be: tex, xdg-utils, flexiblas-devel, tcl-devel, tk-devel.
Btw, the exact same code works perfectly if you start from ubi9 and subscribe to RHEL because all the packages are in RHEL9.
- The other one is definitely more fancy, I would not ask for this to always work. It's a base image with OpenCV and PyTorch built from source using some optimization:  
[ubi9-opencv-pytorch.log](https://github.com/sclorg/s2i-python-container/files/10211125/ubi9-opencv-pytorch.log)
Here the missing packages would be: tbb-devel, libv4l-devel, eigen3-devel, freeglut-devel, mesa-, libGL-devel, boost, boost-thread, boost-devel, gstreamer1-plugins-base-devel, lapack, lapack-devel, lmdb-devel, tesseract-devel
Again, this second case simply illustrates that UBI images are great to install things, but not that much to be used as compiling platforms. Which raises another question: what to use to compile software that will be later installed on UBI9 (with a docker multistage build)? Compile on RHEL and inject in UBI, or compile on CentOS? I guess the two are equivalent in content, or close enough, but then I don't really know how this works from a Licensing perspective...

**Comment by pkubatrh** [comment]
Note that we already had a similar request in the past for centos8 (also R related) #347, which resulted in the request to increase UBI repository content (https://bugzilla.redhat.com/show_bug.cgi?id=1774783) but unfortunately got closed.

@fatherlinux might want to be aware that we got more interest for this.

**Comment by pkubatrh** [comment]
>Which raises another question: what to use to compile software that will be later installed on UBI9 (with a docker multistage build)? Compile on RHEL and inject in UBI, or compile on CentOS? 

Definitely compile on RHEL and inject into UBI. Centos Stream 9 always tracks future RHEL9 minor version, so you might get some incompatibilities if you build on that.

**Comment by guimou** [comment]
Definitely! I'd add myself headers, compilers,... only if and when needed. 
It's easier than starting from s2i-base and copy/maintain the Python boilerplate.
Plus it makes a better image anyway as I don't get npm or nodejs that I don't need.

**Comment by frenzymadness** [comment]
I think we don't have to move it. We can them. We need more testers for the concept of the minimal images and this looks like a good way to get them :smile: Let me take a look at this.

**Comment by frenzymadness** [comment]
Is there any equivalent for ubi-minimal for centos stream? I've found only this repo on quay: https://quay.io/repository/centos/centos?tab=tags But there is no minimal image to build on. Do you know where to raise a question like this?

**Comment by pkubatrh** [comment]
Ah right, that might be an issue :/ I expect Josh Boyer would be able to help out or at least point you to the correct people to talk to.

As a workaround until a proper Centos Stream minimal image we could just make a multistage build and create a minimal-like image ourselves, similarly as we do for micro images in [httpd](https://github.com/sclorg/httpd-container/blob/master/2.4-micro/Dockerfile.c9s) and [nginx](https://github.com/sclorg/nginx-container/blob/master/1.22-micro/Dockerfile.c9s).

**Comment by frenzymadness** [comment]
Centos stream 9 minimal images are available on quay.io so I'm proposing Python 3.9 minimal on top of C9S in #579 . If you can help me with testing, that would be great.

**Comment by frenzymadness** [comment]
`python-39-minimal-c9s` and `python-39-c9s` are available and fresh. Would you have time to test them?

**Comment by guimou** [comment]
Thanks a lot @frenzymadness! It will be difficult this week, but I just booked a whole pm next week to test that.
Quick question though: why did you skip Python 3.10 (I see only 3.9 and 3.11)? Many libraries are not yet fully 3.11-compatible, or have been for a few weeks only if not days.

**Comment by frenzymadness** [comment]
Python 3.10 is not available in Centos Stream or RHEL. We have a container image based on Fedora with Python 3.10 but that will soon be EOL because it's based on Fedora 36 with EOL date in the middle of May this year.

**Comment by frenzymadness** [comment]
So, because we have Python 3.9 and 3.11 available on CS and RHEL, I'm closing this issue now.
